### ISSUE TITLE
Detaching vertices in Fulgora computer

### ISSUE THREAD

**Issue by mbroecheler** [thinkaurelius/titan#862]
For sideEffects, retain/except or when a user adds vertices to a list in a sideEffect, Fulgora should replace such vertex with a detached vertex to conserve memory.

It might also makes sense to simply not allow certain types of sideEffects that require setting a vertex for this reason since they cannot be retrieved after the fact.

**Comment by dalaro** [comment]
This seems like it may require coordination between TP3 and Titan, if we're going to return DetachedVertex instances.  Reattachment of GraphComputer result elements seems to happen here, from debugging and squinting at the code:

https://github.com/tinkerpop/tinkerpop3/blob/3.0.0.M7/gremlin-core/src/main/java/com/tinkerpop/gremlin/process/computer/traversal/step/map/ComputerResultStep.java#L51-L52

and even DetachedVertex implements Attachable and would pass that reattachment condition.  Should we make a vertex class for this that is not Attachable?  What do you think is the best way to go about this?

**Comment by okram** [comment]
DetachedXXX can be used anywhere.

Detaching:

``` java
Vertex detachedVertex = DetachedFactory.detach(titanVertex,true) // where true is "include properties"
Property detachedProperty = DetachedFactory.detach(titanProperty)
VertexProperty detachedVertexProperty = DetachedFactory.detach(titanVertexProperty,true)
```

Attaching:

``` java
Graph titanGraph = TitanFactory.open(...)
Vertex titanVertex = detachedVertex.attach(titanGraph)
```

Traversers (in process -- Gremlin OLAP)

```
Traverser<Vertex> t = TraverserFactory.generateTraverser(fulgoraVertex,1l);

/// traverser code
 /////////////////

    @Override
    public Traverser.Admin<T> detach() {
        this.t = DetachedFactory.detach(this.t, false);
        this.path = DetachedFactory.detach(this.path, true);
        return this;
    }

    @Override
    public Traverser.Admin<T> attach(final Vertex vertex) {
        if (this.t instanceof DetachedElement)
            this.t = (T) ((DetachedElement) this.t).attach(vertex);
        else if (this.t instanceof DetachedProperty)
            this.t = (T) ((DetachedProperty) this.t).attach(vertex);
        // you do not want to attach a path because it will reference graph objects not at the current vertex
        return this;
    }

    /////////////////
```

Perhaps at the end of a round, you go through all the properties of the vertex and:

```
FulgoraVertex.properties().forEach(property -> {
  if(property.value() instanceof FulgoraObject)
     // detach and set
     FulgoraObject.detach()
   if(property.value() instanceof List)
     // walk the list and check and reset
}
```

**Comment by okram** [comment]
Check this -- I propagate the "compute properties" between rounds. Attach them to the vertex, VertexComputer.execute(), then detach the properties (which is a recursive detachment), and the vertex is back to its original pristine state. 

https://github.com/apache/incubator-tinkerpop/blob/master/hadoop-gremlin/src/main/java/org/apache/tinkerpop/gremlin/hadoop/process/computer/spark/SparkExecutor.java#L121-L129

**Comment by dalaro** [comment]
I discussed out-of-band with @okram & @mbroecheler.  The current plan:

After Fulgora executes a VertexProgram on a vertex, inspect its associated VertexState (via FulgoraVertexMemory) for element compute keys of standard TP3 Gremlin side effects that retain references to attached elements (e.g. groupCount).  For each element compute key associated with a standard TP3 side effect, "detach" the associated value.  "detach" here probably really means serialize and deserialize the value, since that should already be well-tested and robust due to preexisting use in TP3's GraphComputer impls.  If there is a cheaper way to detach, for instance by making some easy, stable assumptions about the structure of the values of the EC keys and using DetachedFactory or something, then it's worth considering, but it should be weighed against the complexity and brittleness it might introduce.

The point of this is to reduce reference retention associated with element compute keys of standard Gremlin steps when executing under Fulgora.  It's still possible to shoot oneself in the foot even with this change; for instance, `sideEffect{someList.add(it)}` would still accumulate loads of attached references..

**Comment by dalaro** [comment]
@mbroecheler  I think the plan isn't going to work.  Treating the values associated with standard TP3 Gremlin side effect element compute keys as opaque objects and serializing-then-deserializing them means, in the cases we care about for Fulgora, that we attempt to serialize a PreloadedVertex.  Since the Kryo configuration underlying both GryoWriter (M8 only) and StandardSerializer does not have special behavior for PreloadedVertex, we end up applying Kryo's catchall behavior for serializing an arbitrary class on PreloadedVertex.

We could perhaps introspect the values associated with side effect element compute keys, using a pile of `instanceof`/`isAssignableFrom`, attempting recursive descent into its datastructures, looking for `Vertex` or `Edge` objects and running `DetachedFactory.detach` on them, finally overwriting references non-detached references with detached equivalents in the host datastructure(s).  But that seems rather brittle.

This issue is consuming a lot of time and communication, and I'm really wondering if it's worth it.  Is this something we should try to hack into Fulgora, or might it be tidier to delegate to app?  To make it concrete, consider:

```
// groupCount with default behavior -- retains PreloadedVertex and TitanGraph memory
gremlin> gc = g.V().groupCount().submit(g.compute()).next()
==>v[256]=1
==>v[512]=1
==>v[768]=1
==>v[1024]=1
==>v[1280]=1
==>v[1536]=1
gremlin> gc.keySet().iterator().next().getClass()
==>class com.thinkaurelius.titan.graphdb.vertices.PreloadedVertex

// ... vs ...

// Manually detach to reduce reference retention 
gremlin> gc = g.V().groupCount('m').by{DetachedFactory.detach(it,false)}.cap('m').submit(g.compute()).next()
==>v[256]=1
==>v[512]=1
==>v[768]=1
==>v[1024]=1
==>v[1280]=1
==>v[1536]=1
gremlin> gc.keySet().iterator().next().getClass()
==>class com.tinkerpop.gremlin.structure.util.detached.DetachedVertex
gremlin>
```

**Comment by okram** [comment]
If PreloadedVertex implements tinkerpop.Vertex, then Gryo will "detach" it for you automagically.

**Comment by dalaro** [comment]
@okram does TP3 master's Gryo detachment automagic look into side effect maps like groupCount's, or is that something we need additional logic for?  I would test titan09 against 3.0.0-SNAPSHOT right now, but there is a nontrivial set of compile-time incompatibilities when I try to upgrade from M7 to the current master.

**Comment by okram** [comment]
There is nothing special about "sideEffects". Gryo simply serializes the vertex. If the vertex has properties that are "sideEffects", they are serialized. In other words:

```
v.property('name','marko') 
    and
v.property('m',someGroupCountMap)
```

…just properties…

**Comment by dalaro** [comment]
TP3M8 appears to do what we wanted without any additional changes in Titan.  I don't know why.  I'm not even sure if this ninja fixed elsewhere in Titan, or if it's a consequence of a TP3M7 -> M8 change.  However, Fulgora appears to spit out DetachedVertex in its groupCount map whereas Titan OLTP seems to spit out CacheVertex in the same collection.  There could just be a flaw in my test technique (maybe I'm triggering a detach in my traversal below?)

Closing this for now, since I don't see any next steps on M8, but of course reopen if you spot a mistake.

``` java
gremlin> g = TitanFactory.open('conf/titan-berkeleyje.properties')
15/04/24 00:10:49 INFO configuration.GraphDatabaseConfiguration: Set default timestamp provider MICRO
15/04/24 00:10:49 INFO configuration.GraphDatabaseConfiguration: Generated unique-instance-id=7f0001018479-bluebox1
15/04/24 00:10:49 INFO diskstorage.Backend: Initiated backend operations thread pool of size 8
15/04/24 00:10:49 INFO kcvs.KCVSLog: Loaded unidentified ReadMarker start time Timepoint[1429848649479000 μs] into com.thinkaurelius.titan.diskstorage.log.kcvs.KCVSLog$MessagePuller@6f5bd362
==>standardtitangraph[berkeleyje:/home/dalaro/titan/conf/../db/berkeley]
gremlin> // Hack TP Modern schema to workaround Cardinality change between TP3M7 -> M8
==>true
gremlin> m = g.openManagement()
==>com.thinkaurelius.titan.graphdb.database.management.ManagementSystem@472d0f4
gremlin> m.makePropertyKey('name').dataType(String.class).cardinality(Cardinality.LIST).make()
==>name
gremlin> m.makePropertyKey('lang').dataType(String.class).cardinality(Cardinality.LIST).make()
==>lang
gremlin> m.makePropertyKey('age').dataType(Integer.class).cardinality(Cardinality.LIST).make()
==>age
gremlin> m.commit()
==>null
gremlin> g.io().readGryo('data/tinkerpop-modern.gio')
==>null
gremlin> trav = g.traversal().V().groupCount() ; 1
==>1
gremlin> trav.next().keySet().iterator().next().getClass()
15/04/24 00:10:54 WARN transaction.StandardTitanTx: Query requires iterating over all vertices [()]. For better performance, use indexes
==>class com.thinkaurelius.titan.graphdb.vertices.CacheVertex
gremlin> trav = g.traversal().V().groupCount() ; trav.setEngine(new ComputerTraversalEngine(g.compute()))
==>null
gremlin> trav.next().keySet().iterator().next().getClass()
==>class org.apache.tinkerpop.gremlin.structure.util.detached.DetachedVertex
```
